compiler/src/dotty/tools/dotc/transform/TreeChecker.scala

```diff
@@ -514,4 +515,101 @@ object TreeChecker {
   }.apply(tp0)
 
   private val PrivateErased = allOf(Private, Erased)
+
+  /** Check that the tree only contains legal children trees */
+  def checkTreeNode(tree: untpd.Tree)(implicit ctx: Context): Unit = {
+    def assertNotTypeTree(t: untpd.Tree): Unit = assert(!t.isInstanceOf[untpd.TypeTree], tree)
+
+    // TODO add many more sanity checks
+    tree match {
+//      case Ident(name) =>
+       case Select(qualifier, name) =>
+        // FIXME this assertion fails only in tests/run/scala2trait-lazyval.scala
+        // assertNotTypeTree(qual)
+ //      case This(qual) =>
+//
+//      case Super(qual, mix) =>
+       case Apply(fun, args) =>
+        assertNotTypeTree(fun)
+        for (arg <- args) {
+          assertNotTypeTree(arg)
+        }
+       case TypeApply(fun, args) =>
+        assertNotTypeTree(fun)
+ //      case Literal(const) =>
+//
+//      case New(tpt) =>
+       case Typed(expr, tpt) =>
+        assertNotTypeTree(expr)
+ //      case NamedArg(name, arg) =>
+       case Assign(lhs, rhs) =>
+        assertNotTypeTree(lhs)
+        assertNotTypeTree(rhs)
+       case Block(stats, expr) =>
+        for (stat <- stats) {
+          assertNotTypeTree(stat)
+        }
+        assertNotTypeTree(expr)
+       case If(cond, thenp, elsep) =>
+        assertNotTypeTree(cond)
+        assertNotTypeTree(thenp)
+        assertNotTypeTree(elsep)
+ //      case Closure(env, meth, tpt) =>
+       case Match(selector, cases) =>
+        assertNotTypeTree(selector)
+       case CaseDef(pat, guard, body) =>
+        assertNotTypeTree(guard)
+        assertNotTypeTree(body)
+       case Return(expr, from) =>
+        assertNotTypeTree(expr)
+       case Try(block, handler, finalizer) =>
+        assertNotTypeTree(block)
```

Comment:
should be called `cases`, I copied it from some other pattern match. It is a `List[CaseDef]` therefore as precise as it can be and no checking is required.